Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs

```diff
@@ -91,16 +96,90 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
-        var numOps = Math.Max(1,
-            (int)Math.Min(
-                Math.Floor((double)ev.PlayerPool.Count / _cfg.GetCVar(CCVars.NukeopsPlayersPerOp)), _cfg.GetCVar(CCVars.NukeopsMaxOps)));
-        var ops = new IPlayerSession[numOps];
-        for (var i = 0; i < numOps; i++)
+        // Basically copied verbatim from traitor code
+        var playersPerOperative = _cfg.GetCVar(CCVars.NukeopsPlayersPerOp);
+        var maxOperatives = _cfg.GetCVar(CCVars.NukeopsMaxOps);
+
+        var everyone = new List<IPlayerSession>(ev.PlayerPool);
+        var prefList = new List<IPlayerSession>();
+        var cmdrPrefList = new List<IPlayerSession>();
+        var operatives = new List<IPlayerSession>();
+
+        // The LINQ expression ReSharper keeps suggesting is completely unintelligible so I'm disabling it
+        // ReSharper disable once ForeachCanBeConvertedToQueryUsingAnotherGetEnumerator
+        foreach (var player in everyone)
         {
-            ops[i] = _random.PickAndTake(ev.PlayerPool);
+            if(player.ContentData()?.Mind?.AllRoles.All(role => role is not Job {CanBeAntag: false}) ?? false) continue;
+            if (!ev.Profiles.ContainsKey(player.UserId))
+            {
+                continue;
+            }
+            var profile = ev.Profiles[player.UserId];
+            if (profile.AntagPreferences.Contains(NukeopsPrototypeId))
+            {
+                prefList.Add(player);
+            }
+            if (profile.AntagPreferences.Contains(NukeopsCommanderPrototypeId))
+            {
+                cmdrPrefList.Add(player);
+            }
+        }
+
+        var numNukies = MathHelper.Clamp(ev.PlayerPool.Count / playersPerOperative, 1, maxOperatives);
+
+        for (var i = 0; i < numNukies; i++)
+        {
+            IPlayerSession nukeOp;
+            // Only one commander, so we do it at the start
+            if (i == 0)
+            {
+                if (cmdrPrefList.Count == 0)
+                {
+                    if (prefList.Count == 0)
+                    {
+                        if (everyone.Count == 0)
+                        {
+                            Logger.InfoS("preset", "Insufficient ready players to fill up with nukeops, stopping the selection");
+                            break;
+                        }
+                        nukeOp = _random.PickAndTake(everyone);
+                        Logger.InfoS("preset", "Insufficient preferred nukeop commanders or nukies, picking at random.");
+                    }
+                    else
+                    {
+                        nukeOp = _random.PickAndTake(prefList);
+                        Logger.InfoS("preset", "Insufficient preferred nukeop commanders, picking at random from regular op list.");
+                    }
+                }
+                else
+                {
+                    nukeOp = _random.PickAndTake(cmdrPrefList);
```

Comment:
Shouldn't this remove them from the everyone + prefList as well in case of duplicates?